[jit] DeadCodeEliminator Mark(block) improvement

shinyehtsai opened the pull request:
Summary:
This diff seeks to optimize the DeadCodeEliminator within the mark(block) function.

The primary concept is to prevent redundant traversals of a fully marked block, particularly in the markLoop scenario, if all nodes within a block are marked, we can subsequently mark the block as fully marked.

Test Plan: Existing unittest. Will add new soon

Differential Revision: D73476431




cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

gmagogsfm commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -96,9 +103,9 @@ class DeadCodeEliminator {
   // the block).
   //
   // Returns true iff this marked something we haven't marked before.
```
Please update comments here too.

shinyehtsai replied:
Related comments are updated.

gmagogsfm commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -166,54 +173,77 @@ class DeadCodeEliminator {
+      anyMarked |= markedResult.markUpdated_;
+      // Only take the last iteration of the loop into account.
+      allMarked |= markedResult.fullyMarked_;
     } while (marked);
```
`marked` is never updated. Do you intend to only run the `do-while` loop once? if so, then why do we need a `do-while` loop and why do we do `|=` inside it?

shinyehtsai replied:
Resolved and add comments.

gmagogsfm commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -166,54 +173,77 @@ class DeadCodeEliminator {
-      anyMarked |= marked;
+      auto markedResult = mark(node->blocks().at(0));
+      anyMarked |= markedResult.markUpdated_;
+      // Only take the last iteration of the loop into account.
```
Is this assumption right? Did most of the speedup come from here?

shinyehtsai replied:
Comments are updated and specify which part brings speed up

gmagogsfm commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -166,54 +173,77 @@ class DeadCodeEliminator {
+      auto markedResult = mark(node->blocks().at(0));
+      anyMarked |= markedResult.markUpdated_;
+      // Only take the last iteration of the loop into account.
+      allMarked |= markedResult.fullyMarked_;
```
Why `|=` not `&=`?

gmagogsfm replied:
Also a minor nit, for boolean values it is better to use `&&=`

shinyehtsai replied:
Change to use = to make it fit semantic.

gmagogsfm commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -13,6 +13,15 @@ namespace prim {
 using namespace ::c10::prim;
 }
 
+class MarkResult {
```
Could you add unit tests for this change?

shinyehtsai replied:
Done in the latest version.

gmagogsfm reviewed (commented):
Did you get any benchmarking results on this pass? like how much faster it is than before and how much additional memory is used.

shinyehtsai commented:
> Did you get any benchmarking results on this pass? like how much faster it is than before and how much additional memory is used.

In my latest run on a complicated model, the overall performance difference is around 40%.
106773ms -> 66071 ms (post-PR).

gmagogsfm commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -3050,6 +3050,25 @@ TEST(TestShapeGraphLinting, Basic) {
   }
 }
 
+TEST(DeadCodeEliminatorTest, TestConstructor) {
```
This unit test doesn't seem to be meaningful. Could you add some meaningful test for dce pass? There wasn't comprehensive test before but now complexity is higher, we should have more test cases. You can add more here:
https://github.com/pytorch/pytorch/blob/main/test/jit/test_dce.py

shinyehtsai replied:
a new unittest is added.

davidberard98 reviewed (commented):
if you can add some python tests for more complex dce cases in test_dce.py, that would be great - otherwise LGTM!

shinyehtsai committed:
DeadCodeEliminator Mark(block) improvement (#152348)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152348

**TL;DR**: This diff seeks to optimize the DCE within the mark(block) function.

**Details**
The goal of this PR is to optimize this function:
```
MarkResult mark(Block* block) {
    bool anyMarked = false;
    bool allMarked = true;

    // If all nodes within this block are already marked, we can safely bypass
    // revisiting it. This check is the primary driver of our performance
    // optimization.
    if (alreadyAllMarked_.count(block)) {
      return MarkResult(false, true);
    }

```

The primary concept is to prevent redundant traversals of a fully marked block, particularly in the markLoop scenario, if all nodes within a block are marked, we can subsequently mark the block as fully marked.

Test Plan: Rely on unittest

Reviewed By: davidberard98

Differential Revision: D73476431
